tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -899,9 +901,25 @@ TEST_F(AllgatherOverlapTest, AllgatherBasedPipeliningHostIrImplementation) {
   for_loop_stream->body().push_back(sync_stream);
   hic->pushBackTopLevelExprs(for_loop_stream);
 
-  // The following line is artificial but necessary to make
-  // tva_j->isProducerOf(tvc_j) == true
-  hic->addOutput(tvc_j);
+  // The following line is artificial but necessary to make tva_j_unsqueeze a
+  // consumer of tva_j.
+  //
+  // HostIrEvaluator::handle(ForLoop*) relies on `Val::uses()` to find all
+  // **transitive** consumers of the loop index `j`.  `tva_j_unsqueezed` is a
+  // bit special among all transitive consumers of `j`. It doesn't use `j`
+  // directly but uses `tva_j` which is a TensorView.  TensorView's uses are
+  // built lazily by Fusion::resetTvUses. For efficiency, Fusion::resetTvUses
+  // only fix TensorViews that can reach outputs. Therefore, we add
+  // tva_j_unsqueezed as an output.  Other TensorViews don't need this
+  // treatmenet because they are direct users of `j`, a scalar whose uses are
+  // built eagerly upon registration.
+  //
+  // We could have added `tvc_j` instead as an output, which transitively
+  // consumes `tva_j_unsqueezed`. However, `tvc_j` has two definitions, a Select
+  // and a MatmulOp, and StmtSort::getExprs only traverse via the first
+  // registered definition (i.e. the Select). This sounds like a bug -- I wonder
+  // how nvFuser resets the TensorView uses of a kir::Kernel, also non-SSA.
+  hic->addOutput(tva_j_unsqueezed);
```

Comment:
The introduction of `tva_j_unsqueezed` triggered a weird problem that @samnordmann is probably aware of. I added more explanation and wonder what @naoyam think about this.

Reply:
We could make TensorView live even if no output depends on it for HostIR. Not sure if that would solve the issue, though, as I'm still not entirely clear what the issue is.

Reply:
I am aware we artificially need to add the matmul's output as a fusion output to fix the data dependency, that is why `tvc_j` was added in the first place. However, I was not aware of the other bug you're mentioning -- that we only traverse the first registered producing Expr.

Would the program break if you only let `hic->addOutput(tvc_j);` ?

Reply:
> Would the program break if you only let hic->addOutput(tvc_j);?

Yes as I commented at https://github.com/NVIDIA/Fuser/pull/3531/files#diff-30df6421558f87ef0024b01f11752c35d3d68b80a9e6e0ec0fd49de535acb91aR917

Reply:
Ok but I am not sure to fully understand the reason why it breaks. Even if the visitor only traverses through the first definition, i.e., the `SelectOp`, then `tvc_j` should still be invalidated because the `SelectOp` consumes the index `j`

Reply:
Yes, tvcj will be invalidated but tvaj unsqueeze won't be. As a result it holds always hold the first iteration value